### PR TITLE
Homework 6

### PR DESCRIPTION
**Pull request by BraveLittleTapok**

### CODE REVIEW

File: src/test/java/homework_1/dataproviders/DivideDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package homework_1.dataproviders;
```
Comment (khda91) [inline src/test/java/homework_1/dataproviders/DivideDataProvider.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/dataproviders/MultiplyDataProviders.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package homework_1.dataproviders;
```
Comment (khda91) [inline src/test/java/homework_1/dataproviders/MultiplyDataProviders.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/dataproviders/SubtractDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package homework_1.dataproviders;
```
Comment (khda91) [inline src/test/java/homework_1/dataproviders/SubtractDataProvider.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/dataproviders/SumDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package homework_1.dataproviders;
```
Comment (khda91) [inline src/test/java/homework_1/dataproviders/SumDataProvider.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/tests/AbstractBaseTest.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package homework_1.tests;
```
Comment (khda91) [inline src/test/java/homework_1/tests/AbstractBaseTest.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/tests/Constants.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package homework_1.tests;
```
Comment (khda91) [inline src/test/java/homework_1/tests/Constants.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/tests/DivideTest.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package homework_1.tests;
```
Comment (khda91) [inline src/test/java/homework_1/tests/DivideTest.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/tests/MultiplyTest.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package homework_1.tests;
```
Comment (khda91) [inline src/test/java/homework_1/tests/MultiplyTest.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/tests/SubtractTest.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package homework_1.tests;
```
Comment (khda91) [inline src/test/java/homework_1/tests/SubtractTest.java:1]: Why is HW1 file here?

File: src/test/java/homework_1/tests/SumTest.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package homework_1.tests;
```
Comment (khda91) [inline src/test/java/homework_1/tests/SumTest.java:1]: Why is HW1 file here?

File: src/test/java/homework_6/exercise1/RunTest.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+        @DataProvider
+        public Object[][] features() {
+            return testNGCucumberRunner.provideFeatures();
+        }
+
+        @AfterClass(alwaysRun = true)
+        public void tearDownClass() throws Exception {
+            testNGCucumberRunner.finish();
+        }
```
Comment (khda91) [inline src/test/java/homework_6/exercise1/RunTest.java]: Why do you need these method?

File: src/test/java/homework_6/exercise2/WebDriverSingleton.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    private WebDriver driver;
+
+    public void setDriver(WebDriver driver) {
+        this.driver = driver;
+    }
+
+    public WebDriver getDriver() {
+        return driver;
+    }
+}
```
Comment (khda91) [inline src/test/java/homework_6/exercise2/WebDriverSingleton.java]: Why do you need this code duplication?

File: src/test/java/homework_6/exercise2/hooks/CucumberHooks.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+        WebDriver driver = new ChromeDriver();
+        driver.manage().timeouts().implicitlyWait(10000, TimeUnit.MILLISECONDS);
+        WebDriverSingleton.INSTANCE.setDriver(driver);
+    }
+
+    @After
+    public void tearDown() {
+
+        WebDriverSingleton.INSTANCE.getDriver().close();
+    }
```
Comment (khda91) [inline src/test/java/homework_6/exercise2/hooks/CucumberHooks.java]: Why do you need this code duplication?

File: src/test/java/homework_6/exercise2/steps/AbstractBaseSteps.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+ */
+public abstract class AbstractBaseSteps {
+    protected WebDriver driver;
+    protected static HomePage homePage;
+    protected static User user;
+    protected static HeaderMenu headerMenu;
+    protected static UserTablePage tablePage;
+
+
+}
```
Comment (khda91) [inline src/test/java/homework_6/exercise2/steps/AbstractBaseSteps.java]: Code duplication from the excercise 1 package

File: src/test/java/homework_6/exercise2/steps/ActionSteps.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+import cucumber.api.java.en.And;
+import cucumber.api.java.en.When;
+import homework_6.pages.UserTablePage;
+import homework_6.components.HeaderMenu;
+import homework_6.components.User;
+
+import static homework_6.exercise1.WebDriverSingleton.INSTANCE;
+
+public class ActionSteps extends AbstractBaseSteps {
```
Comment (khda91) [inline src/test/java/homework_6/exercise2/steps/ActionSteps.java]: I assume that this steps could be combined to the on class `ActionSteps`?

File: src/test/java/homework_6/exercise2/steps/AssertionsSteps.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+import cucumber.api.java.en.And;
+import cucumber.api.java.en.Then;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class AssertionsSteps extends AbstractBaseSteps {
```
Comment (khda91) [inline src/test/java/homework_6/exercise2/steps/AssertionsSteps.java]: I assume, that these step could be combined with the steps from the `AssertionsSteps` exercise_1 pacakge

File: src/test/java/homework_6/exercise2/steps/ModalStep.java [hunk 17]
```diff
@@ -0,0 +1,15 @@
+package homework_6.exercise2.steps;
+
+import cucumber.api.java.en.Given;
+import homework_6.pages.HomePage;
+
+import static homework_6.exercise1.WebDriverSingleton.INSTANCE;
+
+public class ModalStep extends AbstractBaseSteps {
```
Comment (khda91) [inline src/test/java/homework_6/exercise2/steps/ModalStep.java]: Code duplication

File: src/test/resources/hw6/exercise1/exercise1.feature [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+  Scenario: Check different pages
+    Given I open EPAM JDI Site
+    Then browser title should be 'Home Page'
+    When I login as 'epam' with password '1234'
+    Then username should be 'PITER CHAILOVSKII' on the Home page
+    And 4 pictures are displayed on the Home Page
+    And 4 text are displayed under pictures on the Home Page
+    And headline above the pictures is displayed on the Home Page
+    And description above the pictures is displayed on the Home Page
```
Comment (khda91) [inline src/test/resources/hw6/exercise1/exercise1.feature]: From the BDD practice it is better to say `should be displayed`, Please do not use `is/are` in the assertions steps

File: src/test/resources/hw6/exercise1/exercise1.feature [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    When I click on 'Service' in the header menu
```
Comment (khda91) [inline src/test/resources/hw6/exercise1/exercise1.feature]: On which page? If you use page name it is better to use them in all steps

**Review by khda91 — COMMENTED** [review]
Please resolve comments

File: src/test/java/homework_6/exercise2/WebDriverSingleton.java (hunk 12 above)
Comment (BraveLittleTapok) [reply src/test/java/homework_6/exercise2/WebDriverSingleton.java]: delete dublicate

File: src/test/java/homework_6/exercise2/steps/AbstractBaseSteps.java (hunk 14 above)
Comment (BraveLittleTapok) [reply src/test/java/homework_6/exercise2/steps/AbstractBaseSteps.java]: delete dublicate

File: src/test/resources/hw6/exercise1/exercise1.feature (hunk 19 above)
Comment (BraveLittleTapok) [reply src/test/resources/hw6/exercise1/exercise1.feature]: fix

File: src/test/resources/hw6/exercise1/exercise1.feature (hunk 18 above)
Comment (BraveLittleTapok) [reply src/test/resources/hw6/exercise1/exercise1.feature]: fix

**Review by khda91 — APPROVED** [review]
Approved. Could be merged into master branch

**Pull request merged**